File: Aula03/CPP/Heranca.cpp
#include <stdio.h>
#include <string.h>

using namespace std;

// Classe mãe (classe base)
class Animal 
{
private:
    int numeroPatas;
    char nome[10];
    char especie[10];
    char som[10];

public:
    Animal(int numeroPatas, char* nome, char* especie, char* som) 
    {
        this->numeroPatas = numeroPatas;
        strcpy(this->nome, nome);
        strcpy(this->especie, especie);
        strcpy(this->som, som);
    }

    void emitirSom()
    {
        printf("O %s faz %s!\n", especie, som);
    }
};

// Classe filha (classe derivada)
class Cachorro : public Animal 
{
public:
    Cachorro() : Animal(4, "Cachorro", "Canino", "Auau") { }
};

// Classe filha (classe derivada)
class Gato : public Animal 
{
public:
    Gato() : Animal(4, "Gato", "Felino", "Miau") { }
};

int main() {
    int tam = 2;
    Animal* animal[2];  // Vetor de ponteiros para a classe Animal
    animal[0] = new Cachorro();
    animal[1] = new Gato();

    for (int i = 0; i < tam; i++)
    {
        animal[i]->emitirSom();
    }
    

    for (int i = 0; i < tam; i++)
    {
        delete animal[i];
    }
    

    return 0;
}

File: Aula03/CS/Heranca.cs
using System;

namespace Programa
{
    // Classe mãe (classe base)
    class Animal {
        int numeroPatas;
        string nome;
        string especie;
        string som;

        public Animal(int numeroPatas, string nome, string especie, string som) {
            this.numeroPatas = numeroPatas;
            this.nome = nome;
            this.especie = especie;
            this.som = som;
        }
        public void EmitirSom() {
            Console.WriteLine("{0} faz {1}", nome, som);
        }
    }

    // Classe filha (classe derivada)
    class Cachorro : Animal {
        public Cachorro() : base(4, "Cachorro", "Canis lupus familiaris", "Au au!") { }
    }

    // Classe filha (classe derivada)
    class Gato : Animal {
        public Gato() : base(4, "Gato", "Felis catus", "Miau!") { }
    }

    class MainClass {
        static void Main(string[] args) {
            Animal[] animals = {
                new Animal(6, "Formiga", "Formiga", "Nada"),
                new Cachorro(),
                new Gato()};

            foreach(Animal animal in animals) {
                animal.EmitirSom();
            }

            Console.ReadLine();
        }
    }
}